src/streaming-client.ts

```diff
@@ -0,0 +1,157 @@
+     */
+    constructor(accessToken: string, config: AudioConfig, version = 'v1alpha') {
+        super();
+        this.accessToken = accessToken;
+        this.config = config;
+        this.baseUrl = `wss://api.rev.ai/speechtotext/${version}/stream?` +
+            `access_token=${accessToken}` +
+            `&content_type=${config.getContentTypeString()}`;
+        this.requests = new PassThrough();
+        this.responses = new PassThrough({ objectMode: true });
+        this.client = new sockets.client();
+    }
+
+    /**
+     * Sets up the client and begins the streaming connection. Returns a duplex
+     * from which the user can read responses from the api and to which the user
+     * should write their audio data
+     *
+     * @returns BufferedDuplex. Data written to this buffer will be sent to the api
+     * Data received from the api can be read from this duplex
+     */
+    start(): Duplex {
+        this.setUpClient();
+        this.client.connect(this.baseUrl);
+        return new BufferedDuplex(this.requests, this.responses, { readableObjectMode: true });
+    }
+
+    /**
+     * Ends the streaming connection and closes off the buffer returned from start()
+     */
+    end(): void {
+        this.client.abort();
+        this.requests.end('EOS');
+        this.responses.push(null);
+    }
+
+    private setUpClient(): void {
+        this.setUpHttpResponseHandler();
+        this.setUpConnectionFailureHandler();
+        this.setUpConnectedHandlers();
+    }
+
+    private setUpHttpResponseHandler(): void {
+        this.client.on('httpResponse', (response: any) => {
+            this.emit('httpResponse', response.statusCode);
+            this.end();
+        });
+    }
+
+    private setUpConnectionFailureHandler(): void {
+        this.client.on('connectFailed', (error: Error) => {
+            this.emit('connectFailed', error);
+            this.end();
+        });
+    }
+
+    private setUpConnectedHandlers(): void {
+        this.client.on('connect', (connection: any) => {
+            connection.on('error', (error: Error) => {
+                this.emit('error', error);
+                this.end();
+            });
+            connection.on('close', (code: number, reason: string) => {
+                this.emit('close', code, reason);
+                this.end();
+            });
+            connection.on('message', (message: any) => {
+                if (message.type === 'utf8') {
+                    let response = JSON.parse(message.utf8Data);
+                    if ((response as StreamingResponse).type === 'connected') {
+                        this.emit('connect', response as StreamingConnected);
+                    } else {
+                        this.responses.write(response as StreamingHypothesis);
+                    }
+                }
+            });
+
+            function sendFromBuffer(buffer: PassThrough): void {
```

Comment:
you can use arrow functions to define all instance functions

moving the function definition out of the on 'connect' callback will make this a bit cleaner.

sendFromBuffer = (buffer: PassThrough): Void => {
}